### PR TITLE
Alexandra Chakeres

### PR DESCRIPTION
**Pull request by chakeresa**
Allllmost done with iteration 4 (the patrons_of_exhibits hash is a bit messed up at the moment -- not passing the full integration test)

### CODE REVIEW

**Comment by chakeresa** [comment]
Just kidding, I thought Bob should be under imax instead of dead sea scrolls, but he shouldn't. My "expected" variable was wrong but the code is working as expected. :-)

File: lib/exhibit.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class Exhibit
+  attr_reader :name, :cost
+  
+  def initialize(name, cost)
+    @name = name
+    @cost = cost
+  end
+end
```
Comment (mstang15) [inline lib/exhibit.rb]: 💯

File: lib/patron.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class Patron
+  attr_reader :name, :interests
+
+  attr_accessor :spending_money
```
Comment (mstang15) [inline lib/patron.rb]: Nice call here!

File: test/exhibit_test.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  def test_it_exists
+    assert_instance_of Exhibit, @exhibit
+  end
+
+  def test_it_inits_with_name_and_cost
+    assert_equal "Gems and Minerals", @exhibit.name
+    assert_equal 0, @exhibit.cost
+  end
+end
```
Comment (mstang15) [inline test/exhibit_test.rb]: 💯

File: test/patron_test.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    assert_equal [], @bob.interests
+  end
+
+  def test_add_interest_adds_to_interests_ary
+    @bob.add_interest("Dead Sea Scrolls")
+    @bob.add_interest("Gems and Minerals")
+
+    assert_equal ["Dead Sea Scrolls", "Gems and Minerals"], @bob.interests
+  end
+end
```
Comment (mstang15) [inline test/patron_test.rb]: 💯

File: test/museum_test.rb [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    expected = {
+      gems_and_minerals => [morgan], # bob shouldn't be attending (but isn't) :-(
+      imax => [bob, sally], # bob should be attending (but isn't) :-(
+      dead_sea_scrolls => [morgan]
+    }
+
+    assert_equal expected, dmns.patrons_of_exhibits
+    assert_equal 35, dmns.revenue
+  end
+end
```
Comment (mstang15) [inline test/museum_test.rb]: Great job on this test!

File: lib/museum.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  def add_exhibit(exhibit)
+    @exhibits << exhibit
+  end
+
+  def recommend_exhibits(patron)
+    @exhibits.find_all do |exhibit|
+      patron.interests.include?(exhibit.name)
+    end
+  end
```
Comment (mstang15) [inline lib/museum.rb]: Great use of #find_all and #include?

File: lib/museum.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+  def admit(patron)
+    @patrons << patron
+    sorted_exhibits = recommend_exhibits(patron).sort_by do |exhibit|
```
Comment (mstang15) [inline lib/museum.rb]: I like that you're reusing your recommend_exhibits method here! Nice job

File: lib/museum.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  def patrons_by_exhibit_interest
+    @exhibits.inject(Hash.new([])) do |hash, exhibit|
+      interested_patrons = @patrons.find_all do |patron|
+        patron.interests.include?(exhibit.name)
+      end
+      hash[exhibit] = interested_patrons
+      hash
+    end
```
Comment (mstang15) [inline lib/museum.rb]: Great job here!

File: lib/museum.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+              :patrons,
+              :revenue,
+              :patrons_of_exhibits
+
+  def initialize(name)
+    @name = name
+    @exhibits = []
+    @patrons = []
+    @revenue = 0
+    @patrons_of_exhibits = {}
```
Comment (mstang15) [inline lib/museum.rb]: If you initialized @patrons_of_exhibits = Hash.new([]) Then you could get rid of your code on lines 46-48

File: lib/museum.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  def attend(patron, exhibit)
+    cost = exhibit.cost
+    patron.spending_money -= cost
+    @revenue += cost
+    if !@patrons_of_exhibits.keys.include?(exhibit)
+      patrons_of_exhibits[exhibit] = []
+    end
+    @patrons_of_exhibits[exhibit] << patron
+    @patrons_of_exhibits
+  end
```
Comment (mstang15) [inline lib/museum.rb]: All of this code looks like it should work. I even pulled it down and played around with it. I think that there may be an assertion in your test that is wrong. Maybe spend some time picking that apart?

**Comment by chakeresa** [comment]
@mstang15 Yeah I figured out after time was up that my expected hash was wrong for that last test. The code was doing exactly what it was supposed to! I just admitted Bob into an IMAX movie he couldn't afford, haha! I fixed it now.

**Pull request closed without merging**